File: src/TypeFormatter/RestTypeFormatter.ts
import { Definition } from "../Schema/Definition.js";
import { SubTypeFormatter } from "../SubTypeFormatter.js";
import { BaseType } from "../Type/BaseType.js";
import { RestType } from "../Type/RestType.js";
import { TypeFormatter } from "../TypeFormatter.js";

export class RestTypeFormatter implements SubTypeFormatter {
    public constructor(protected childTypeFormatter: TypeFormatter) {}

    public supportsType(type: BaseType): boolean {
        return type instanceof RestType;
    }

    public getDefinition(type: RestType): Definition {
        const definition = this.childTypeFormatter.getDefinition(type.getType());
        const title = type.getTitle();

        if (title !== null && typeof definition.items === "object") {
            return { ...definition, items: { ...definition.items, title } };
        }

        return definition;
    }

    public getChildren(type: RestType): BaseType[] {
        return this.childTypeFormatter.getChildren(type.getType());
    }
}

File: internal/formatter/rest_type_formatter.go
package formatter

import (
	"github.com/vega/ts-json-schema-generator/internal/schema"
	"github.com/vega/ts-json-schema-generator/internal/types"
)

// RestTypeFormatter mirrors src/TypeFormatter/RestTypeFormatter.ts.
type RestTypeFormatter struct {
	childTypeFormatter TypeFormatter
}

func NewRestTypeFormatter(childTypeFormatter TypeFormatter) *RestTypeFormatter {
	return &RestTypeFormatter{childTypeFormatter: childTypeFormatter}
}

func (f *RestTypeFormatter) SupportsType(t types.Type) bool {
	_, ok := t.(*types.RestType)
	return ok
}

func (f *RestTypeFormatter) GetDefinition(t types.Type) *schema.Definition {
	restType := t.(*types.RestType)
	definition := f.childTypeFormatter.GetDefinition(restType.Type)
	title := restType.Title

	if title != "" {
		if items, ok := definition.Items.(*schema.Definition); ok {
			clone := definition.Clone()
			itemsClone := items.Clone()
			itemsClone.Title = title
			clone.Items = itemsClone
			return clone
		}
	}

	return definition
}

func (f *RestTypeFormatter) GetChildren(t types.Type) []types.Type {
	return f.childTypeFormatter.GetChildren(t.(*types.RestType).Type)
}
